test/model/internal_link_test.dart

```diff
@@ -0,0 +1,512 @@
+      (false, 'wrong domain',
+        Uri.parse('https://another.com/#narrow/stream/check'), realmUrl),
+
+      (false, '#narrowly',
+        realmUrl.resolve('#narrowly/stream/check'), realmUrl),
+
+      (false, 'double slash',
+        Uri.parse('https://example.com//#narrow/stream/check'), realmUrl),
+      (false, 'triple slash',
+        Uri.parse('https://example.com///#narrow/stream/check'), realmUrl),
+
+      (true, 'with port',
+        Uri.parse('https://example.com:444/#narrow/stream/check'),
+        Uri.parse('https://example.com:444')),
+
+      // Uri.parse performs percent-encoding of unicode domains:
+      // `Uri.parse('https://example.भारत/')` has an origin of
+      // `https://example.%E0%A4%AD%E0%A4%BE%E0%A4%B0%E0%A4%A4`
+      // so in the following examples unicode domains will
+      // pass but punycode variants of unicode domains are not
+      // treated as matching.
+
+      // (true, 'same domain, punycoded host',
+      //   Uri.parse('https://example.xn--h2brj9c/#narrow/stream/check'),
+      //   Uri.parse('https://example.भारत/')), // FAILS
+
+      (true, 'punycodable host',
+        Uri.parse('https://example.भारत/#narrow/stream/check'),
+        Uri.parse('https://example.भारत/')),
+
+      // (true, 'same domain, IDNA-mappable',
+      //   Uri.parse('https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check'),
+      //   Uri.parse('https://example.com')), // FAILS
+
+      (true, 'ipv4 address',
+        Uri.parse('http://192.168.0.1/#narrow/stream/check'),
+        Uri.parse('http://192.168.0.1/')),
+
+      // (true, 'same IPv4 address, IDNA-mappable',
+      //   Uri.parse('http://１𝟗𝟚。①⁶🯸．₀｡𝟭/#narrow/stream/check'),
+      //   Uri.parse('http://192.168.0.1/')), // FAILS
+
+      // TODO: Add tests for IPv6.
+
+      // These examples may seem weird, but a previous version of
+      // the zulip-mobile code accepted most of them.
+
+      // This one, except possibly the fragment, is a 100% realistic link
+      // for innocent normal use.  The buggy old version narrowly avoided
+      // accepting it... but would accept all the variations below.
+      (false, 'wrong domain, realm-like path, narrow-like fragment',
+        Uri.parse('https://web.archive.org/web/*/${realmUrl.resolve('#narrow/stream/check').toString()}'),
+        realmUrl),
+      (false, 'odd scheme, wrong domain, realm-like path, narrow-like fragment',
+        Uri.parse('ftp://web.archive.org/web/*/${realmUrl.resolve('#narrow/stream/check').toString()}'),
+        realmUrl),
+      (false, 'same domain, realm-like path, narrow-like fragment',
+        realmUrl.resolve('web/*/${realmUrl.resolve('#narrow/stream/check').toString()}'),
+        realmUrl),
+    ];
+    for (final testCase in testCases) {
+      final bool expected = testCase.$1;
+      final String because = testCase.$2;
+      final Uri url = testCase.$3;
+      final Uri realmUrl = testCase.$4;
+      test(because, () async {
```

Comment:
The zulip-mobile version has a bit more in the test description:
```js
    test(`${expected ? 'accept' : 'reject'} ${description}: ${url.toString()}`, () => {
```

I think the accept/reject part, especially, is helpful when you're looking at a list of these that failed, or passed, and trying to understand what they're saying.